AsyncNats/NatsConnection.cs

```diff
@@ -157,32 +165,67 @@ private async Task ReadWriteAsync(CancellationToken disconnectToken)
 
                 try
                 {
-                    _logger?.LogTrace("Connecting to {Server}", Options.Server);
-                    await socket.ConnectAsync(Options.Server);
+                    _currentServer = _serverPool.SelectServer(_isRetry);
+                    _logger?.LogTrace("Connecting to {Server}", _currentServer);
+
+                    IPEndPoint ipEndpoint;
+                    if (IPAddress.TryParse(_currentServer.Host, out var ipAddress))
+                    {
+                        ipEndpoint = new IPEndPoint(ipAddress, _currentServer.Port);
+                    }
+                    else
+                    {
+                        _logger?.LogTrace("Resolving dns hostname {Host}", _currentServer.Host);
+
+                        try
+                        {
+                            var resolvedHost = await (Options.DnsResolver ?? Dns.GetHostAddressesAsync)(_currentServer.Host);
+
+                            var selectedIpAddress = resolvedHost.OrderBy(i => Guid.NewGuid()).First();
+                            _logger?.LogTrace("Dns hostname {Host} resolve to {Ip}", _currentServer.Host, selectedIpAddress);
+                            ipEndpoint = new IPEndPoint(selectedIpAddress, _currentServer.Port);
+                        }
+                        catch(SocketException ex)
+                        {
+                            _logger?.LogWarning("Failed to resolve dns hostname {Host}", _currentServer.Host);
+                            throw;
+                        }
+                        
+                    }
+
+                    await socket.ConnectAsync(ipEndpoint);
                 }
                 catch (Exception ex)
                 {
-                    _logger?.LogError(ex, "Error connecting to {Server}", Options.Server);
+                    _logger?.LogError(ex, "Error connecting to {Server}", _currentServer);
                     ConnectionException?.Invoke(this, ex);
 
                     await Task.Delay(TimeSpan.FromSeconds(1), disconnectToken);
+
+                    _isRetry = true;
+
                     continue;
                 }
 
+                _isRetry = false;
                 _receiverQueueSize = 0;
                 
                 var readPipe = new Pipe(new PipeOptions(pauseWriterThreshold: 1024*1024));
                 // ReSharper disable AccessToDisposedClosure
-                var readTask = Task.Run(() => ReadSocketAsync(socket, readPipe.Writer, internalDisconnect.Token), internalDisconnect.Token);
-                var processTask = Task.Run(() => ProcessMessagesAsync(readPipe.Reader, internalDisconnect.Token), internalDisconnect.Token);
-                var writeTask = Task.Run(() => WriteSocketAsync(socket, internalDisconnect.Token), internalDisconnect.Token);
+
+                var combinedCts = CancellationTokenSource.CreateLinkedTokenSource(disconnectToken, internalDisconnect.Token);
```

Comment:
I'm missing a dispose?     
````C#
using var combinedCts = CancellationTokenSource.CreateLinkedTokenSource(disconnectToken, internalDisconnect.Token);
````
This will probably work.

Reply:
True